src/agent/src/watcher.rs

```diff
@@ -0,0 +1,487 @@
+}
+
+impl Entry {
+    async fn new(storage: Storage) -> Result<Entry> {
+        let source = PathBuf::from(&storage.source);
+        let mut mount_point = PathBuf::from(&storage.mount_point);
+
+        if source.is_file() && mount_point.is_dir() {
+            let filename = source.file_name().with_context(|| {
+                format!("Failed to extract file name from {}", source.display())
+            })?;
+            mount_point = mount_point.join(filename);
+        }
+
+        let mut entry = Entry::default();
+        entry.add_path(&source).await?;
+
+        entry.source = source;
+        entry.mount_point = mount_point;
+
+        Ok(entry)
+    }
+
+    /// Adds file entries from the given path
+    #[async_recursion]
+    async fn add_path(&mut self, path: &Path) -> Result<()> {
+        if path.is_file() {
+            self.add_file(path)
+                .await
+                .with_context(|| format!("Failed to add file {}", path.display()))?;
+        } else {
+            let mut entries = fs::read_dir(path)
+                .await
+                .with_context(|| format!("Failed to read dir: {}", path.display()))?;
+
+            while let Some(entry) = entries.next_entry().await? {
+                self.add_path(entry.path().as_path()).await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Adds a single file path entry
+    async fn add_file(&mut self, file: &Path) -> Result<()> {
+        ensure!(
+            self.files.len() <= MAX_ENTRIES_PER_STORAGE,
+            "Too many file system entries to watch (must be < {})",
+            MAX_ENTRIES_PER_STORAGE
+        );
+
+        let entry = File {
+            file: file.to_path_buf(),
+            // Set minimum value to perform initial copy during next check
+            modified: SystemTime::UNIX_EPOCH,
+        };
+
+        self.files.push(entry);
+        Ok(())
+    }
+
+    async fn check(&mut self) -> Result<usize> {
+        let mut count = 0;
+        for i in 0..self.files.len() {
+            if self.check_file(i).await? {
+                count += 1;
+            }
+        }
+
+        Ok(count)
+    }
+
+    async fn check_file(&mut self, i: usize) -> Result<bool> {
+        let entry = &mut self.files[i];
+
+        let current = modified(&entry.file).await?;
+        if current <= entry.modified {
+            return Ok(false);
+        }
+
+        // Update entry
+        entry.modified = current;
+
+        // We no longer need mutable borrow, reborrow.
+        let entry = &self.files[i];
+
+        // File to file copy
+        if self.source.is_file() {
+            fs::copy(&self.source, &self.mount_point).await?;
+            return Ok(true);
+        }
+
+        // Dir to dir copy
+        let source_file_path = entry.file.as_path();
+        let dest_file_path = self.make_dest_path(&source_file_path)?;
+
+        if let Some(path) = dest_file_path.parent() {
+            fs::create_dir_all(path).await?;
```

Comment:
The dest_file_path parent dir would exist, thus it should check the case and skip to create it again, otherwise, it would cause an error here.